File: src/contest/hackerrank/WOC_28_D.cpp
#include <bits/stdc++.h>

#define mp make_pair
#define pb push_back
#define INF 1<<30
#define MOD 1000000007
#define rint(x) scanf("%d", &(x))
#define rlong(x) scanf("%lld", &(x))
#define MAX_N 100000
#define l(x) x << 1
#define r(x) x << 1|1
#define m(x, y) (x + y)/2

using namespace std;

typedef long long ll;
typedef pair<int, int> pi;
typedef pair<ll, ll> pll;

int Q, N, M;
vector<vector<int>> adj;
bool vis[MAX_N];
long calc[MAX_N + 1];

int bfs (int n) {
    queue<int> q;
    q.push(n);
    vis[n] = true;
    int ret = 1;
    while (!q.empty()) {
        int u = q.front();
        q.pop();
        for (int v : adj[u])
            if (!vis[v]) {
                vis[v] = true;
                q.push(v);
                ret++;
            }

    }
    return ret;
}

int main () {
	// freopen("in.txt", "r", stdin);
	// freopen("out.txt", "w", stdout);

    scanf("%d", &Q);

    for (int i = 1; i <= MAX_N; i++)
        calc[i] = (ll)(i) * (i - 1) + calc[i - 1];

    for (int q = 0; q < Q; q++) {
        scanf("%d%d", &N, &M);

        adj.clear();
        ll ans = 0;

        memset(vis, 0, sizeof(vis));

        for (int i = 0; i < N; i++)
            adj.push_back(vector<int>());

        for (int i = 0; i < M; i++) {
            int u, v;
            scanf("%d%d", &u, &v);
            u--, v--;
            adj[u].push_back(v);
            adj[v].push_back(u);
        }

        vector<int> sizes;
        for (int i = 0; i < N; i++)
            if (!vis[i]) {
                int sz = bfs(i);
                if (sz > 1)
                    sizes.push_back(sz);
            }

        sort(sizes.begin(), sizes.end());
        reverse(sizes.begin(), sizes.end());

        ll edgesLeft = M;
        for (int i = 0; i < (int)sizes.size(); i++) {
            edgesLeft -= sizes[i] - 1;
            ans += (edgesLeft * sizes[i] * (sizes[i] - 1)) + calc[sizes[i]];
        }

        printf("%lld\n", ans);
    }

	return 0;
}

File: src/contest/hackerrank/WOC_28_D.java
package contest.hackerrank;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Queue;
import java.util.StringTokenizer;

public class WOC_28_D {

  static final int MAX_N = 100000;
  static BufferedReader br;
  static PrintWriter out;
  static StringTokenizer st;
  static int Q, N, M;
  static ArrayList<ArrayList<Integer>> adj = new ArrayList<ArrayList<Integer>>();
  static boolean[] vis;
  static long[] calc = new long[MAX_N + 1];

  public static void main(String[] args) throws IOException {
    br = new BufferedReader(new InputStreamReader(System.in));
    out = new PrintWriter(new OutputStreamWriter(System.out));
    //br = new BufferedReader(new FileReader("in.txt"));
    //out = new PrintWriter(new FileWriter("out.txt"));

    Q = readInt();

    for (int i = 1; i <= MAX_N; i++)
      calc[i] = (long) (i) * (i - 1) + calc[i - 1];

    for (int q = 0; q < Q; q++) {
      N = readInt();
      M = readInt();

      adj.clear();
      vis = new boolean[N];
      long ans = 0;

      for (int i = 0; i < N; i++)
        adj.add(new ArrayList<Integer>());

      for (int i = 0; i < M; i++) {
        int u = readInt() - 1;
        int v = readInt() - 1;
        adj.get(u).add(v);
        adj.get(v).add(u);
      }

      ArrayList<Integer> sizes = new ArrayList<Integer>();
      for (int i = 0; i < N; i++) {
        if (!vis[i]) {
          int sz = fill(i);
          if (sz > 1)
            sizes.add(sz);
        }
      }

      Collections.sort(sizes);
      Collections.reverse(sizes);

      long edgesLeft = M;
      for (int i = 0; i < sizes.size(); i++) {
        edgesLeft -= (sizes.get(i) - 1);
        ans += (edgesLeft * sizes.get(i) * (sizes.get(i) - 1)) + calc[sizes.get(i)];
      }

      out.println(ans);
    }

    out.close();
  }

  static int fill(int u) {
    Queue<Integer> q = new ArrayDeque<Integer>();
    q.offer(u);
    vis[u] = true;
    int ret = 1;
    while (!q.isEmpty()) {
      int curr = q.poll();
      for (int v : adj.get(curr))
        if (!vis[v]) {
          vis[v] = true;
          q.offer(v);
          ret++;
        }
    }
    return ret;
  }

  static String next() throws IOException {
    while (st == null || !st.hasMoreTokens())
      st = new StringTokenizer(br.readLine().trim());
    return st.nextToken();
  }

  static long readLong() throws IOException {
    return Long.parseLong(next());
  }

  static int readInt() throws IOException {
    return Integer.parseInt(next());
  }

  static double readDouble() throws IOException {
    return Double.parseDouble(next());
  }

  static char readCharacter() throws IOException {
    return next().charAt(0);
  }

  static String readLine() throws IOException {
    return br.readLine().trim();
  }
}
